### ISSUE TITLE
JSON format token properties error silently when used by other properties.

### ISSUE THREAD

**Issue by Akaku-Keninari** [RPTools/maptool#2396]
**Describe the bug**
Token properties in JSON format attempt to parse, fail silently, then are treated as strings correctly when referenced by other token properties.

**To Reproduce**
1. Create a token with a property named `firstProperty` that is `{"A": 1, "B": 2}`
2. Create a displayable property on the same token with a value of `{json.get(firstProperty, "A")}`
3. Mouse over the token and see that it displays "1"
4. Check debug log, you should see errors like this:
```
2021-02-18 23:56:34.086 DEBUG net.rptools.maptool.model.Token - Evaluating property: 'Elevation' for token EXAMPLE(4F71C395E5004F778BFB58D9D1357590)----------------------------------------------------------------------------------
2021-02-18 23:56:34.090 DEBUG net.rptools.maptool.client.MapToolLineParser - json.get(Strength, "A")
2021-02-18 23:56:34.105 DEBUG net.rptools.maptool.model.Token - Evaluating property: 'Strength' for token EXAMPLE(4F71C395E5004F778BFB58D9D1357590)----------------------------------------------------------------------------------
2021-02-18 23:56:34.105 DEBUG net.rptools.maptool.client.MapToolLineParser -  "A": 1, "B": 2
2021-02-18 23:56:34.107 DEBUG net.rptools.maptool.client.MapToolLineParser - net.rptools.parser.ParserException: line 1:4: unexpected char: ':'
```

**Expected behavior**
It shouldn't spend the time to error out silently. This time adds up on frameworks that have extreme amounts of jsons and calls to other properties in their statsheets.

**MapTool Info**
- Version: 1.8.3
- Install: New

**Desktop (please complete the following information):**
 - OS: Windows
 - Version 10

**Additional context**
Attached is zip file that's actually a .cmpgn file set up already.

[json error.zip](https://github.com/RPTools/maptool/files/6008477/json.error.zip)

Link to a campaign that slows to a crawl because of this:
https://drive.google.com/file/d/1Z3iGg0oKnAJV6vcCxmpoAmhC7BTiiuha/view

Getting the json with `getRawProperty()` bypasses this error but has limitations (will not pull defaults) and is messy to hack into existing code.
Messy hack to 'fix' the issue on a token in linked campaign:
`[h: names = getAllPropertyNames(getPropertyType(),"json")]
[foreach(prop, names, "<br>"), code:
{
	[r: prop]
	[h: rawp = if ( getRawProperty(prop) == "", getPropertyDefault(prop) , getRawProperty(prop))]
	[h: newstring = replace(rawp, '(?:json\\.get\\()([A-Za-z]*)(?:,)', 'json\\.get\\(getRawProperty\\("\$1"\\),')]
	[h: setProperty(prop, newstring)]
}]
`

**Comment by Azhrei** [comment]
I don’t understand.  If malformed JSONs are accessed slowly (as strings) when malformed, isn’t the best solution to ensure they aren’t malformed when stored into the property in the first place?  The contract between MT and the user is that properties that start with a valid JSON character (bracket or brace) are, in fact, JSON values.  MT is currently obligated to verify this when retrieving values.

I say “currently” because some future version of MT could assign data types to token properties and/or variables such that attempts to store improperly formatted data would trigger an error at that time, instead of when the value is retrieved.

(I’ve not tried replying from my phone before. Apologies for any formatting weirdness.)
--
Frank Edwards
Edwards & Edwards Consulting, LLC
Hours:  10A-6P ET
Phone:  (813) 406-0604

**Comment by Akaku-Keninari** [comment]
The problem is that they are not malformed. When accessing it with just the name from another property, it checks if it's runable code, fails, then treats it as a string.

**Comment by Azhrei** [comment]
> 1. Create a token with a property named `firstProperty` that is `{"A": 1, "B": 2"}`

That’s malformed.  There is a missing double-quote.

**Comment by Akaku-Keninari** [comment]
> 
> 
> > 1. Create a token with a property named `firstProperty` that is `{"A": 1, "B": 2"}`
> 
> That’s malformed. There is a missing double-quote.

I made a typeo in the report. It does the issue when it's `{"A": 1, "B": 2}`

**Comment by Pmofmalasia** [comment]
Thanks Akaku for looking into this! I'm the writer of the campaign that's getting slowed down by this issue, and I should clarify a few extra details.

I've since tested it out in 1.7 and this error was not present with the exact same campaign file.

I also initially thought it was only a problem when referenced by other token properties on the statsheet, but it appears to actually occur any time a token property that is a JSON is referenced. Normally in macros they aren't referenced often enough for the amount of time spent to be noticeable, but the "Check" and "Save" macros did so often enough that the delay was noticeable.

For ease of use, I have uploaded a version of the campaign with Akaku's fix applied to most tokens, as well as a troubleshooting map that contains a token with macros that can be used to show the effects outside of the statsheet.

https://drive.google.com/file/d/18xzmnWS1ykkMoZJrCh8BiCwaGjHp0Qjp/view

**Comment by selquest** [comment]
Haven't looked at this in detail yet, but at first glance it looks like it's trying to parse as MTscript first, THEN deciding it's json.  Hazzard of supporting syntax like:
<pre>
[h: myVar = 7]
{myVar}
</pre>

**Comment by Pmofmalasia** [comment]
Agreed, especially since the error message displayed is the same as would be shown in a non-token property macro if `[h:{"a":0}]` was done. I tried working around this issue by changing the format of json objects to `[h:'{"a":0}']` or `[r:'{"a":0}']` but that just results in the error
> Argument number 2 "a" to function "json.get" must be an integer.

which I believe is because of changes to how jsons are handled in 1.8.3 from what I was reading in the macros discord channel. 
`[r:json.set("","a",0)]` technically functions correctly but is much more difficult to update the value.

**Comment by selquest** [comment]
Looked some more at the history of this, and as expected this is basically the result of #1560.  We used try and treat it as JSON first, but the Gson JSON parsers are much more forgiving of syntax, and have a tendency to evaluate things like `{a1=b2}` as a valid JSON object... and since it was a valid JSON object, parsing as MTScript didn't continue.  To fix that problem, the order of evaluations of token properties now is:
1. null/empty string
1. JSON Array
2. MTScript expression
3. JSON Object
4. Number
5. String

I would expect the MTScript parser to be substantially slower than the JSON evaluators, so there is definitely a performance cost to doing things in this order, especially if JSON Objects are more common in properties than MTScript expressions are.  I think the ideal order of evaluation would have MTScript as the very last assumption, but again the problem comes down to overlapping syntax.  If we can't convince the JSON evaluators to more effectively defeat MTScript expressions, the only way to do so is to actually try and parse them.

**Comment by selquest** [comment]
Given the performance hit that we're getting from this evaluation order, it may be worth exploring alternative solutions to #1560.  Akaiku provided a link with guidance that looks potentially helpful in making the Gson parsers less lenient, though I haven't investigated further: https://stackoverflow.com/questions/43233898/how-to-check-if-json-is-valid-in-java-using-gson/47890960#47890960

**Comment by selquest** [comment]
I'm planning to experiment with this a bit, but I'm also now looking at the provided campaign in more detail.  I use a similar framework in my own games (they have a common ancestor in Ao's 5e framework), and definitely haven't been experiencing the same kind of terrible slowdowns apparent here.  I added some performance timers on the statsheet popup and found the following performance numbers when trying to load the statsheet for the reference Cleric on the Library map (times in ms):

Timer | Provided | My Own
---|--:|--:
**Total** | 12,594 | 336
dAC | 3,064 | 50
dPassiveInvestigation | 252 | 27
dPassivePerception | 262 | 26
dResistance | 1,265 | 16
dSpeed | 2,630 | 15

Several other properties have much closer evaluation times, but I'm going to dig into the differences in the definitions of these properties that have such wild performance differences.

**Comment by Pmofmalasia** [comment]
This may be partially due to my addition of an expanded magic item system on top of Ao's framework, which uses a massive object with 378 keys (MagicItemStats) that is then called by other stats on the statsheet. To use the variables you gave, with the number of times they call this object:

dAC: 6
dPassiveInvestigation: 0
dPassivePerception: 0
dResistance: 39
dSpeed: 5

Note that the numbers from dAC, dResistance, and dSpeed are actually from AC, Resistance, and Speed since they reference those properties. Speed and AC show the most obvious correlation with calls to MagicItemStats, but Resistance completely demolishes that relationship - possibly because its calls to MagicItemStats are in a UDF? The most surprising to me is that PassiveInv/Perc are increased even though they're completely unchanged from Ao's framework.

**Comment by Pmofmalasia** [comment]
On a separate note, how do you anticipate handling the order of evaluations for future updates, while alternative solutions to #1560 are investigated? Obviously this is significantly affecting my framework but I understand that mine seems to be an edge case, and many more people are likely to want to use `{var1=var2}` format than will be affected by this issue. On the other hand, from what I can tell the previous order was the standard since 1.5.12, and so hopefully reverting won't affect too many campaigns - and the issue is solved much easier by switching to `[r:var1=var2]` (unless there is a compelling reason not to do so which I am unaware of).

Basically, I can see an argument for sticking with either, I'd just like to know so I can adjust my framework accordingly - my leading solution at the moment is to just stick all of my objects into an array, since they seem unaffected.

**Comment by selquest** [comment]
It does appear that the majority of extra time spent in your campaign is coming from the repeated attempts to parse `MagicItemStats` - and most of that time seems to be coming from the repeated application of the regex transformer.  Not terribly surprising, but I still want to see if I can find a way to avoid adding literal seconds there.

Also did a quick test restoring the original order (basically reverting #1560 locally), and as expected that dropped the overall processing time of the above test down to 780ms.  I'm going to experiment using the strict parsing approach linked above to see if this is viable - while I personally can't think of any good reason that `{var1 = var2}` should be used instead of just defining `var1` as `{var2}` in the first place, the syntax is technically valid and we need to make sure it doesn't produce an actual wrong outcome... but especially as we've been steering people to use more and more json lately we have to do it in a way that doesn't cripple that usage like this.

Some discussion with Craig on discord brought up an approach that's been discussed before, and is probably better as a long-term solution: add some caching to property resolution, so that literals can be retrieved immediately without having to go through the whole parsing process.

In the meantime, for a solution that fits better in a bugfix release, I'm going to try and pursue the strict json evaluation.  To maintain backwards compatibility with existing frameworks that might be taking advantage of our historically lenient evaluation, my goal is to change the above evaluation order:
1. null/empty string
2. JSON array
3. JSON object, strict evaluation
4. MTScript parse
5. JSON object, lenient fall-back
6. Number
7. String

Hopefully this is achievable as a kind of middle ground that can be included in the next release.

**Comment by Pmofmalasia** [comment]
Great, that sounds like a good middle ground! By any chance, would reverting the processing order be something that I could feasibly do on my end? It would be good to have a temporary fix between now and the next update without having to restructure my code - and I assume my players would have to do so as well?

> Some discussion with Craig on discord brought up an approach that's been discussed before, and is probably better as a long-term solution: add some caching to property resolution, so that literals can be retrieved immediately without having to go through the whole parsing process.

This would be amazing! As you can see I'm pushing the limit of what will feel responsive on the character sheet, so that would help a lot.

**Comment by selquest** [comment]
> Great, that sounds like a good middle ground! By any chance, would reverting the processing order be something that I could feasibly do on my end? It would be good to have a temporary fix between now and the next update without having to restructure my code - and I assume my players would have to do so as well?

If you and your players are comfortable producing/running unofficial versions whose compatibility with released versions would obviously not be guaranteed, there's no particular reason you wouldn't be able to... but that's obviously not a recommended approach 😉

Since 1.8.4 just went out it's likely to be at least a few weeks before another release is pushed, so even if we get this "fixed" you'll have a bit of an interim need.  My suggestion for a workaround for now would be to wrap MagicItemStats in an Array as you mentioned before.  You may get even better performance converting the definitions that contain multiple references into UDFs that could unpack MagicItemStats once into a local variable instead of needing to re-evaluate with each reference.

Of course, given that MagicItemStats appears to just be an arbitrary collection of independent named values (*cough* also known as a List of Properties) you could also simplify an awful lot of this by just splitting that monolithic JSON object into separate properties:
```
MagicItemStats.bAC: 0
MagicItemStats.sAC: 0
MagicItemStats.soAC: -1
(etc)
```

**Comment by Pmofmalasia** [comment]
> Of course, given that MagicItemStats appears to just be an arbitrary collection of independent named values (cough also known as a List of Properties)

Ha! Believe me, this is making me wish I had done that before. At the time I figured it would be neater to squirrel them all away in one property to avoid scrolling through ~400 properties if anything needed editing. Though, it does bring up something I had been curious about - is there a theoretical maximum number of token properties? I imagine that would be heavily dependent on the complexity of those properties, of course.

**Comment by selquest** [comment]
> is there a theoretical maximum number of token properties? I imagine that would be heavily dependent on the complexity of those properties?

So it's funny you should ask that, because that's the second part of this I'm working on now.  There's no actual defined maximum number of properties, so you're limited only by your memory considerations (and I guess some of the data structures used might be limited to java's integer maximum of around 2 billion).  That said, the list of property definitions is currently implemented as a `List` that is iterated across to find a definition that matches the given property name.  That means the more properties you have the less performant it is to lookup the default value.

I'm going to replace that List with a LinkedHashMap to allow O(1) lookups while maintaining the definition order.  In practice though, the performance gain here is not going to be particularly noticeable outside of contexts with at least hundreds (maybe thousands) of properties.

**Comment by selquest** [comment]
Changed my mind about "part 2" being included here, because it's a bigger change with more related touchpoints and deserves its own issue (#2484) that is likely better deferred to 1.9 than included in the next bugfix release.

The existing PR is ready, and wraps up work on this issue.

**Comment by Phergus** [comment]
Tested.  No longer spams the log file.
